### DISCUSSION TITLE
Integrating Form with Combobox Custom Options

### DISCUSSION THREAD

**Discussion by mekhu** [mantinedev/mantine#5815] [Q&A]
Hello, I would like to ask if there's anyone who can provide an example on integrating Mantine forms to the Combobox with Custom Options? I'm following the example here: https://mantine.dev/combobox/?e=SelectOptionComponent and I just added some auto complete to it. I can't seem to understand how to use the `form.getInputProps('value')`. Here's my current implementation:

`

    import { useState } from 'react';
    import { Avatar, Combobox, Group, Input, InputBase, Text, useCombobox } from '@mantine/core';
      
      interface MemberPicker {
          id: string;
          name: string;
          image: string;
      }
      
      interface IComponentModel {
          value?: string;
          onChange?(event: React.ChangeEvent<HTMLInputElement>): void;
          onFocus?(event: React.FocusEvent<HTMLInputElement>): void;
          onBlur?(event: React.FocusEvent<HTMLInputElement>): void;
          error?: string;
      }
      
      const groceries: MemberPicker[] = [
          { id: "2b929334-aeaa-4866-802e-75a14f4427a1", image: "https://i.pravatar.cc/300?img=1", name: "Member Name 1" },
          { id: "b6ce0a2d-08ab-4a29-8116-5d4e1fe52a0e", image: "https://i.pravatar.cc/300?img=2", name: "Member Name 2" },
          { id: "9af5b6df-d649-40c7-87e3-d629fb1a7af8", image: "https://i.pravatar.cc/300?img=3", name: "Member Name 3" },
          { id: "dbab7ca1-3f9e-410e-9eab-5c53e12f5a23", image: "https://i.pravatar.cc/300?img=4", name: "Member Name 4" },
          { id: "4022588b-aab9-4911-b71c-55094720ead5", image: "https://i.pravatar.cc/300?img=5", name: "Member Name 5" },   
      ];
      
      function SelectOption({ id, image, name }: MemberPicker) {
          return (
              <Group gap="xs">
                  <Avatar src={image} alt="no image here" radius="xl" />
                  <div>
                      <Text fz="sm" fw={500}>
                          {name}
                      </Text>
                  </div>
              </Group>
          );
      }
      
      export function MembersPicker({ value, onChange, onFocus, onBlur, error }: IComponentModel) {
          const [val, setVal] = useState<string | null>(null);
          const [search, setSearch] = useState('');
      
          const combobox = useCombobox({
              onDropdownClose: () => {
                  combobox.resetSelectedOption(),
                      combobox.focusTarget();
                  setSearch('');
              },
              onDropdownOpen: () => {
                  combobox.focusSearchInput();
              },
          });
      
      
          const options = groceries
              .filter((item) => item.name.toLowerCase().includes(search.toLowerCase().trim()))
              .map((item) => (
                  <Combobox.Option value={item.id} key={item.id}>
                      <SelectOption {...item} />
                  </Combobox.Option>
              ));
      
      
          const selectedOption = groceries.find((item) => item.id === val);
      
          return (
              <Combobox
                  store={combobox}
                  withinPortal={false}
                  onOptionSubmit={(val) => {
                      setVal(val);
                      combobox.closeDropdown();
                  }}
              >
                  <Combobox.Target>
                      <InputBase
                          component="button"
                          type="button"
                          pointer
                          rightSection={<Combobox.Chevron />}
                          onClick={() => combobox.toggleDropdown()}
                          rightSectionPointerEvents="none"
                          multiline
                      >
                          {selectedOption ? (
                              <SelectOption {...selectedOption} />
                          ) : (
                              <Input.Placeholder>Pick value</Input.Placeholder>
                          )}
                      </InputBase>
                  </Combobox.Target>
      
                  <Combobox.Dropdown>
                      <Combobox.Search
                          value={search}
                          onChange={(event) => setSearch(event.currentTarget.value)}
                          placeholder="Search members"
                      />
                      <Combobox.Options mah={200} style={{ overflowY: 'auto' }}>
                          {options.length > 0 ? options : <Combobox.Empty>Nothing found</Combobox.Empty>}
                      </Combobox.Options>
                  </Combobox.Dropdown>
              </Combobox>
          );
      }

`

Any guidance or suggestions would be greatly appreciated. Thank you!

**Comment by rtivital** [comment]
You can follow this guide with dropzone, the logic should be similar – https://help.mantine.dev/q/how-to-use-dropzone-with-form

**Reply by mekhu** [reply]
Is this applicable if I would like to use this custom combobox as a reusable component? How can I pass the getInputProps inside the custom combobox component and consume it in the code? Thank you

**Reply by rtivital** [reply]
You custom combobox component must support `onChange` and `value` props, then you will be able to spread `form.getInputProps()` to it the same way as any other input. Optionally, you can add `error`, `onFocus` and `onBlur` props handling in your component to support all form features.

**Reply by mekhu** [reply]
Thank you. But based on my current code, where can I place the "value" and "onchange" props?

**Reply by lux4198** [reply]
I am facing the same problem, where I have a custom combobox component, I want to pass `{...form.getInputProps("")}` to like with the other inputs in the form. The only way I got it to work, was to pass the entire form as props and then call `form.setFieldValue()` manually in the Combobox `onOptionSubmit` prop. Is this the intended way to do this?

**Reply by laugri** [reply]
@lux4198 I have the exact same use case, I haven't found another way to do this with Combobox since the combobox value isn't triggered by `onChange` but by `onOptionSubmit`. 😕

**Reply by mekhu** [reply]
Here's what I've done:

	import { useEffect, useState } from 'react';
	import { CloseButton, Combobox, Input, InputBase, useCombobox } from '@mantine/core';
	import { MemberPicker, MemberPickerOption } from '.';

	interface IComponentModel {
		onSelect: (member?:MemberPicker) => void;
		error?: string;
		[props: string]: any;
		value?: string;
	};

	const membersList: MemberPicker[] = [
		{ id: "2b929334-aeaa-4866-802e-75a14f4427a1", image: "https://i.pravatar.cc/300?img=1", name: "Member Name 1" },
		{ id: "b6ce0a2d-08ab-4a29-8116-5d4e1fe52a0e", image: "https://i.pravatar.cc/300?img=2", name: "Member Name 2" },
		{ id: "9af5b6df-d649-40c7-87e3-d629fb1a7af8", image: "https://i.pravatar.cc/300?img=3", name: "Member Name 3" },
		{ id: "dbab7ca1-3f9e-410e-9eab-5c53e12f5a23", image: "https://i.pravatar.cc/300?img=4", name: "Member Name 4" },
		{ id: "4022588b-aab9-4911-b71c-55094720ead5", image: "https://i.pravatar.cc/300?img=5", name: "Member Name 5" },
	];

	export const MembersPicker = ({ onSelect, error, value, ...props }: IComponentModel) => {
		const [selectedOption, setSelectedOption] = useState<MemberPicker | undefined>();
		const [search, setSearch] = useState('');

		const combobox = useCombobox({
			onDropdownClose: () => {
				combobox.resetSelectedOption(),
					combobox.focusTarget();
				setSearch('');
			},
			onDropdownOpen: () => {
				combobox.focusSearchInput();
			},
		});

		const onMemberSelect = (member?: MemberPicker) => {
			if(member?.id===selectedOption?.id){
				return;
			}
			 onSelect(member);
			 setSelectedOption(member);
		};

		useEffect(() => {
			if(!selectedOption){
				const member = membersList.find((item) => item.id === value);
				setSelectedOption(member);
			}
			if(!value){
				setSelectedOption(undefined);
			} 
		}, [value]);

		const options = membersList
			.filter((item) => item.name.toLowerCase().includes(search.toLowerCase().trim()))
			.map((item) => (
				<Combobox.Option value={item.id} key={item.id}>
					<MemberPickerOption member={item} onMemberSelect={onMemberSelect} />
				</Combobox.Option>
			));

		return (
			<Combobox
				store={combobox}
				withinPortal={false}
				onOptionSubmit={() => {
					combobox.closeDropdown();
				}}>
				<Combobox.Target>
					<InputBase
						component="button"
						type="button"
						pointer
						rightSection={
							selectedOption !== null ? (
								<CloseButton
									size="sm"
									onClick={() => onMemberSelect()}
									aria-label="Clear value"
								/>
							) : (
								<Combobox.Chevron />
							)
						}
						onClick={() => combobox.toggleDropdown()}
						rightSectionPointerEvents={selectedOption === null ? 'none' : 'all'}
						multiline
						error={error}
						{...props}
					>
						{selectedOption ? (
							<MemberPickerOption member={selectedOption} onMemberSelect={onMemberSelect} />
						) : (
							<Input.Placeholder>{props?.placeholder}</Input.Placeholder>
						)}
					</InputBase>
				</Combobox.Target>

				<Combobox.Dropdown>
					<Combobox.Search
						value={search}
						onChange={(event) => setSearch(event.currentTarget.value)}
						placeholder="Search members"
					/>
					<Combobox.Options mah={200} style={{ overflowY: 'auto' }}>
						{options.length > 0 ? options : <Combobox.Empty>Nothing found</Combobox.Empty>}
					</Combobox.Options>
				</Combobox.Dropdown>
			</Combobox>
		);
	};

And implemented it as the following:

	const onMemberSelect = (member?: MemberPicker) => {
			form.setFieldValue('invitorID', member?.id ?? null);
			form.setFieldValue('invitorName', member?.name ?? null);
		};
	
	<MembersPicker
                label="Who invited you? Leave empty if none"
                placeholder="Select an invitor"
                mb="sm" onSelect={onMemberSelect} {...form.getInputProps('invitorID')} />


Here's a screenshot of the component:
![image](https://github.com/mantinedev/mantine/assets/15007956/2f0f9ea9-cd75-451a-9df1-54fafe8be8d4)

Hope this helps!
